lib/auth/iam_join.go

```diff
@@ -0,0 +1,332 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/gravitational/trace"
+	"google.golang.org/grpc/peer"
+)
+
+const (
+	expectedSTSIdentityRequestBody = "Action=GetCallerIdentity&Version=2011-06-15"
+	stsHost                        = "sts.amazonaws.com"
```

Comment:
Hmm, thanks for bringing this up. To quote the docs you linked here:

> By default, the AWS Security Token Service (AWS STS) is available as a global service, and all AWS STS requests go to a single endpoint at https://sts.amazonaws.com. AWS recommends using Regional AWS STS endpoints instead of the global endpoint to reduce latency, build in redundancy, and increase session token validity.

Since we are generating the request on the Teleport node we have complete control over which region and endpoint will be used. So I believe what I have here will work, but it would be possible to optimize for latency and redundancy by using regional endpoints.

If we were to try to optimize this by choosing a region, it's not clear which would be best. If optimizing for redundancy, we should probably use region of the node, but if we wanted to optimize for latency we should use the region of the auth server (which we probably don't know). Also, multiple STS regions such as Milan, Bahrain, Cape Town, and Hong Kong are disabled by default so we would need to figure out what to do then.

All this to say: I think what I have here will work, and trying to optimize now by using regional endpoints would introduce more complexity and ultimately make this feature more fragile. If we are concerned about having the global AWS STS endpoint as a single point of failure and want to provide the option to use regional endpoints, we can always revisit this. Let me know if you agree

Reply:
The AWS SDK STS call from node is based on `SharedConfigState`  
```
 sess, err := session.NewSessionWithOptions(session.Options{
		SharedConfigState: session.SharedConfigEnable,
	})
```
Thus if default AWS region was changed or `AWS_REGION` env was the default sts endpoint might not be used. 
Not sure how AWS credentials will be propagated to the node but if we are sure that the default region won't be overwritten then seems to be ok.

Reply:
Just tried testing this by setting the region in the shared config file and the env var, it will still use the global sts endpoint unless you explicitly request the regional endpoint with `stsService := sts.New(sess, aws.NewConfig().WithSTSRegionalEndpoint(endpoints.RegionalSTSEndpoint))`

> Not sure how AWS credentials will be propagated to the node

The purpose of this feature is to allow nodes to join the cluster without a secret token if they already have AWS credentials. So we are not propagating any AWS credentials, the node should already have them. Since any EC2 instance with an attached IAM role (or a AWS fargate pod with an execution role) should already have credentials this should be easier than provisioning a secret token and getting it to the node for those cases.